cmd/livepeer/livepeer.go

```diff
@@ -225,7 +230,33 @@ func main() {
 		}
 	}
 
+	drivers.DataDir = *datadir
+
+	if *s3bucket != "" && *s3creds == "" || *s3bucket == "" && *s3creds != "" {
+		glog.Error("Should specify both s3bucket and s3creds")
+		return
+	}
+	if *s3bucket != "" {
+		s3bp := strings.Split(*s3bucket, "/")
+		drivers.S3REGION = s3bp[0]
+		drivers.S3BUCKET = s3bp[1]
+	}
+
+	// XXX get s3 credentials from local env vars?
+	if *s3bucket != "" && *s3creds != "" {
+		br := strings.Split(*s3bucket, "/")
+		cr := strings.Split(*s3creds, "/")
+		drivers.ExternalStorage = drivers.NewOwnS3Driver(br[0], br[1], cr[0], cr[1])
+	}
 	if n.NodeType == core.Broadcaster {
+		drivers.LocalSaveMode = *localSave
+		if drivers.ExternalStorage != nil {
+			drivers.AddStorageInstance(drivers.ExternalStorage)
+		}
+		if drivers.LocalSaveMode || drivers.ExternalStorage == nil {
+			drivers.LocalStorage = drivers.NewLocalDriver("")
```

Comment:
no, this one is for broadcaster. for transcoder we initialize it later (actually earlier in the control flow) because for transcoder we need to get it's public URL

Reply:
Isn't the same logic true for the transcoder? If the transcoder has S3 then that should be used, rather than local storage?

Reply:
T creates both S3 and local instances, what will be used depends on what B sends with each segment.

Reply:
Not sure about that. This is the flow I was thinking of:

```
If B has S3
  T uses B's S3
else if T has S3
  T uses T's S3
else
  T uses local storage
```

The last two cases can be collapsed into one -- `T.Storage = if T has S3 then T.S3 else Local`

In fact, the abstraction can probably carry over for B as well:

```
B.Storage = SegmentInfo.Storage // could be B's S3, IPFS, etc
Job.Storage = nil
if B.Storage != nil
  Job.Storage = B.Storage
else
  Job.Storage = T.Storage
...
Job.Storage.Save(...)
```

In other words, I'm not sure that there is a case *right now* where T would have both its own local storage and S3? I can understand supporting multiple storage backends someday for redundancy, but as mentioned, we don't have a clean way to express that via the networking protocol right now.

Reply:
Right now I'm always saving data into T's local storage. But you're right, it can be removed if T's S3 is used.